com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JdtUtils.java

```diff
@@ -78,4 +98,164 @@ public static IJavaProject getJavaProject(String projectName) {
         IProject project = root.getProject(projectName);
         return getJavaProject(project);
     }
+
+    /**
+     * Given the project name, return the corresponding project object.
+     * If the project doesn't exist, return null.
+     */
+    public static IProject getProject(String projectName) {
+        if (projectName == null) {
+            return null;
+        }
+        IWorkspaceRoot root = ResourcesPlugin.getWorkspace().getRoot();
+        return root.getProject(projectName);
+    }
+
+    /**
+     * Compute the possible source containers that the current project could be associated with.
+     * <p>
+     * If current project is specified, try to parse the source containers from current project's classpath entries first,
+     * then the other projects at the same workspace.
+     * </p>
+     * <p>
+     * Otherwise, loop every projects at the current workspace and combine the parsed source containers directly.
+     * </p>
+     * @param currentProject
+     *                  the current project name.
+     * @return the possible source container list.
+     */
+    public static ISourceContainer[] getSourceContainers(String currentProject) {
+        Set<ISourceContainer> containers = new LinkedHashSet<>();
+        List<IProject> projects = new ArrayList<>();
+
+        IProject targetProject = JdtUtils.getProject(currentProject);
+        if (targetProject != null) {
+            projects.add(targetProject);
+        }
+
+        List<IProject> workspaceProjects = Arrays.asList(ResourcesPlugin.getWorkspace().getRoot().getProjects());
+        projects.addAll(workspaceProjects.stream().filter((project) -> {
+            return !project.equals(targetProject);
+        }).collect(Collectors.toList()));
+
+        Set<IRuntimeClasspathEntry> calculated = new LinkedHashSet<>();
+        for (IProject project : projects) {
+            IJavaProject javaProject = JdtUtils.getJavaProject(project);
+            if (javaProject != null && project.exists()) {
+                // Add source containers associated with the project's runtime classpath entries.
+                containers.addAll(Arrays.asList(getSourceContainers(javaProject, calculated)));
+                // Add source containers associated with the project's source folders.
+                containers.add(new JavaProjectSourceContainer(javaProject));
+            }
+        }
+
+        return containers.toArray(new ISourceContainer[0]);
+    }
+
+    private static ISourceContainer[] getSourceContainers(IJavaProject project, Set<IRuntimeClasspathEntry> calculated) {
+        if (project != null && project.exists()) {
```

Comment:
Use early return.

if (project == null || !project.exists()) {
     return *empty*;
}

In this way we avoid indenting large chunk of code.

Reply:
nice.